File: cortex/ide/adapters/vscode.py
from __future__ import annotations

import contextlib
import json
import logging
from pathlib import Path
from typing import Any

from cortex.ide.base import (
    IDEAdapter,
    _backup_file,
    _deep_merge_dict,
    _generate_autogen_header,
)
from cortex.ide.prompts import get_subagent_prompt, strip_markdown_frontmatter

logger = logging.getLogger(__name__)

# Agents canonicos que este adapter escribe (SSoT para install y uninstall).
_VSCODE_TOP_AGENTS: tuple[str, ...] = ("cortex-sync", "cortex-SDDwork")
_CLAUDE_SUBAGENTS: tuple[str, ...] = (
    "cortex-code-explorer",
    "cortex-code-implementer",
    "cortex-documenter",
)


def _render_vscode_agent(frontmatter: list[str], header: str, body: str) -> str:
    frontmatter_block = "\n".join(frontmatter)
    return f"---\n{frontmatter_block}\n---\n\n<!--\n{header.strip()}\n-->\n\n{body.strip()}\n"


def _render_claude_agent(name: str, description: str, header: str, body: str) -> str:
    return (
        f"---\n"
        f"name: {name}\n"
        f"description: {description}\n"
        f"---\n\n"
        f"<!--\n{header.strip()}\n-->\n\n"
        f"{body.strip()}\n"
    )


class VSCodeAdapter(IDEAdapter):
    @property
    def name(self) -> str:
        return "vscode"

    @property
    def display_name(self) -> str:
        return "VS Code"

    def get_config_paths(self) -> dict[str, Path]:
        return {
            "mcp": Path(".vscode") / "mcp.json",
            "agents_dir": Path(".github") / "agents",
            "claude_agents_dir": Path(".claude") / "agents",
        }

    def inject_profiles(self, project_root: Path, prompts: dict[str, str]) -> list[str]:
        paths = self.get_config_paths()
        agents_dir = project_root / paths["agents_dir"]
        claude_agents_dir = project_root / paths["claude_agents_dir"]
        agents_dir.mkdir(parents=True, exist_ok=True)
        claude_agents_dir.mkdir(parents=True, exist_ok=True)

        sync_body = strip_markdown_frontmatter(prompts.get("cortex-sync", ""))
        work_body = strip_markdown_frontmatter(prompts.get("cortex-SDDwork", ""))

        top_level_header = _generate_autogen_header(
            sources=[".cortex/skills/cortex-sync.md", ".cortex/skills/cortex-SDDwork.md"],
            ide_name="vscode",
        )
        subagent_sources = {
            "cortex-code-explorer": (
                "Read-only architecture analysis for complex changes.",
                ".cortex/subagents/cortex-code-explorer.md",
            ),
            "cortex-code-implementer": (
                "Deep-track implementation specialist for complex changes.",
                ".cortex/subagents/cortex-code-implementer.md",
            ),
            "cortex-documenter": (
                "Session documentation and vault persistence specialist.",
                ".cortex/subagents/cortex-documenter.md",
            ),
        }

        files_written: list[str] = []

        sync_path = agents_dir / "cortex-sync.agent.md"
        _backup_file(sync_path)
        sync_path.write_text(
            _render_vscode_agent(
                [
                    "name: cortex-sync",
                    "description: Create Cortex specs before implementation.",
                    "tools: ['search/codebase', 'search/usages', 'cortex/*']",
                    "handoffs:",
                    "  - label: Continue with cortex-SDDwork",
                    "    agent: cortex-SDDwork",
                    "    prompt: Continue from the persisted Cortex spec and execute the implementation workflow.",
                    "    send: false",
                ],
                top_level_header,
                sync_body,
            ),
            encoding="utf-8",
        )
        files_written.append(str(sync_path))

        work_path = agents_dir / "cortex-SDDwork.agent.md"
        _backup_file(work_path)
        work_path.write_text(
            _render_vscode_agent(
                [
                    "name: cortex-SDDwork",
                    "description: Implement Cortex specs with fast-track or deep-track routing.",
                    "tools: ['agent', 'edit', 'search/codebase', 'search/usages', 'cortex/*']",
                    "agents: ['cortex-code-explorer', 'cortex-code-implementer', 'cortex-documenter']",
                ],
                top_level_header,
                work_body,
            ),
            encoding="utf-8",
        )
        files_written.append(str(work_path))

        for agent_name, (description, source) in subagent_sources.items():
            agent_header = _generate_autogen_header(sources=[source], ide_name="vscode")
            agent_path = claude_agents_dir / f"{agent_name}.md"
            _backup_file(agent_path)
            agent_path.write_text(
                _render_claude_agent(
                    agent_name,
                    description,
                    agent_header,
                    strip_markdown_frontmatter(get_subagent_prompt(project_root, agent_name)),
                ),
                encoding="utf-8",
            )
            files_written.append(str(agent_path))

        return files_written

    def inject_mcp(self, project_root: Path) -> list[str]:
        mcp_path = project_root / self.get_config_paths()["mcp"]
        mcp_path.parent.mkdir(parents=True, exist_ok=True)
        _backup_file(mcp_path)

        data: dict[str, Any] = {}
        if mcp_path.exists():
            with contextlib.suppress(Exception):
                data = json.loads(mcp_path.read_text(encoding="utf-8"))

        cortex_config = {
            "type": "stdio",
            "command": "cortex",
            "args": ["mcp-server", "--stdio", "--project-root", "${workspaceFolder}"],
            "env": {"PYTHONWARNINGS": "ignore"},
        }

        data.setdefault("servers", {})
        data["servers"] = _deep_merge_dict(data["servers"], {"cortex": cortex_config})

        mcp_path.write_text(json.dumps(data, indent=2), encoding="utf-8")
        return [str(mcp_path)]

    def uninstall(self, project_root: Path | None = None) -> list[str]:
        """Eliminar lo inyectado por Cortex en VS Code:

        - ``.github/agents/{cortex-sync,cortex-SDDwork}.agent.md``.
        - ``.claude/agents/{explorer,implementer,documenter}.md``.
        - Entrada ``servers.cortex`` de ``<project>/.vscode/mcp.json``.

        SOLO se borra lo que Cortex creo (archivos por nombre canonico);
        los directorios solo se eliminan si quedan vacios. Sin
        ``project_root`` no se toca nada y se emite un warning explicito
        (nunca ``Path.cwd()``).
        """
        if project_root is None:
            logger.warning(
                "[Cortex][VSCode] uninstall() llamado sin project_root: "
                "no-op. Pasa el project root explicito para desinstalar."
            )
            return []

        removed: list[str] = []
        root = Path(project_root).resolve()

        for dir_key, names in (
            ("agents_dir", _VSCODE_TOP_AGENTS),
            ("claude_agents_dir", _CLAUDE_SUBAGENTS),
        ):
            agents_dir = root / self.get_config_paths()[dir_key]
            for agent_name in names:
                suffix = ".agent.md" if dir_key == "agents_dir" else ".md"
                agent_path = agents_dir / f"{agent_name}{suffix}"
                if agent_path.exists():
                    agent_path.unlink()
                    removed.append(str(agent_path))
            if agents_dir.exists() and not any(agents_dir.iterdir()):
                agents_dir.rmdir()
                removed.append(str(agents_dir))

        # Entrada cortex de .vscode/mcp.json (merge inverso, preservando
        # otros servers del adopter).
        mcp_path = root / self.get_config_paths()["mcp"]
        if mcp_path.exists():
            with contextlib.suppress(Exception):
                data = json.loads(mcp_path.read_text(encoding="utf-8"))
                servers = data.get("servers")
                if isinstance(servers, dict) and "cortex" in servers:
                    del servers["cortex"]
                    mcp_path.write_text(
                        json.dumps(data, indent=2), encoding="utf-8"
                    )
                    removed.append(f"{mcp_path} (cortex entry removed)")

        return removed

File: rust/crates/cortex-setup/src/ide/adapters/vscode.rs
//! Porteo de cortex/ide/adapters/vscode.py (P8d).
//!
//! VS Code: agents top-level nativos en `.github/agents/*.agent.md` (con
//! frontmatter tools/handoffs) + subagents en `.claude/agents/*.md` +
//! MCP en `.vscode/mcp.json` bajo la clave `servers` con
//! `${workspaceFolder}` como project-root.

use std::path::PathBuf;

use serde_json::{json, Value};

use crate::ide::base::{backup_file, deep_merge_dict, generate_autogen_header, json_dump_ascii};
use crate::ide::prompts::{get_subagent_prompt, strip_markdown_frontmatter};
use crate::ide::{IdeAdapter, IdeCtx, Prompts};

/// Agents canónicos top-level que este adapter escribe (SSoT install/uninstall).
const VSCODE_TOP_AGENTS: [&str; 2] = ["cortex-sync", "cortex-SDDwork"];
const CLAUDE_SUBAGENTS: [&str; 3] = [
    "cortex-code-explorer",
    "cortex-code-implementer",
    "cortex-documenter",
];

fn render_vscode_agent(frontmatter: &[&str], header: &str, body: &str) -> String {
    let frontmatter_block = frontmatter.join("\n");
    format!(
        "---\n{frontmatter_block}\n---\n\n<!--\n{}\n-->\n\n{}\n",
        header.trim(),
        body.trim()
    )
}

fn render_claude_agent(name: &str, description: &str, header: &str, body: &str) -> String {
    format!(
        "---\nname: {name}\ndescription: {description}\n---\n\n<!--\n{}\n-->\n\n{}\n",
        header.trim(),
        body.trim()
    )
}

pub struct VSCodeAdapter;

impl VSCodeAdapter {
    #[allow(dead_code)]
    pub fn new() -> Self {
        VSCodeAdapter
    }
}

impl Default for VSCodeAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl IdeAdapter for VSCodeAdapter {
    fn name(&self) -> &'static str {
        "vscode"
    }

    fn display_name(&self) -> &'static str {
        "VS Code"
    }

    /// Rutas RELATIVAS al proyecto (Python devuelve Paths relativos; los
    /// call-sites hacen el join con project_root).
    fn config_paths(&self, _ctx: &IdeCtx) -> Vec<(String, PathBuf)> {
        vec![
            ("mcp".into(), PathBuf::from(".vscode").join("mcp.json")),
            ("agents_dir".into(), PathBuf::from(".github").join("agents")),
            (
                "claude_agents_dir".into(),
                PathBuf::from(".claude").join("agents"),
            ),
        ]
    }

    fn inject_profiles(&self, ctx: &IdeCtx, prompts: &Prompts) -> Result<Vec<String>, String> {
        let root = ctx.project_root;
        let agents_dir = root.join(".github").join("agents");
        let claude_agents_dir = root.join(".claude").join("agents");
        std::fs::create_dir_all(&agents_dir)
            .map_err(|e| format!("mkdir {}: {e}", agents_dir.display()))?;
        std::fs::create_dir_all(&claude_agents_dir)
            .map_err(|e| format!("mkdir {}: {e}", claude_agents_dir.display()))?;

        let sync_body = strip_markdown_frontmatter(
            prompts.get("cortex-sync").map(|s| s.as_str()).unwrap_or(""),
        );
        let work_body = strip_markdown_frontmatter(
            prompts
                .get("cortex-SDDwork")
                .map(|s| s.as_str())
                .unwrap_or(""),
        );

        let top_level_header = generate_autogen_header(
            ctx,
            &[
                ".cortex/skills/cortex-sync.md",
                ".cortex/skills/cortex-SDDwork.md",
            ],
            "vscode",
        );

        // Orden de inserción del dict de Python: explorer, implementer,
        // documenter. (La description del documenter difiere de la del
        // adapter claude_code — así está en la fuente vscode.py.)
        let subagent_sources: [(&str, &str, &str); 3] = [
            (
                "cortex-code-explorer",
                "Read-only architecture analysis for complex changes.",
                ".cortex/subagents/cortex-code-explorer.md",
            ),
            (
                "cortex-code-implementer",
                "Deep-track implementation specialist for complex changes.",
                ".cortex/subagents/cortex-code-implementer.md",
            ),
            (
                "cortex-documenter",
                "Session documentation and vault persistence specialist.",
                ".cortex/subagents/cortex-documenter.md",
            ),
        ];

        let mut files_written: Vec<String> = Vec::new();

        let sync_path = agents_dir.join("cortex-sync.agent.md");
        backup_file(ctx, &sync_path);
        let content = render_vscode_agent(
            &[
                "name: cortex-sync",
                "description: Create Cortex specs before implementation.",
                "tools: ['search/codebase', 'search/usages', 'cortex/*']",
                "handoffs:",
                "  - label: Continue with cortex-SDDwork",
                "    agent: cortex-SDDwork",
                "    prompt: Continue from the persisted Cortex spec and execute the implementation workflow.",
                "    send: false",
            ],
            &top_level_header,
            &sync_body,
        );
        std::fs::write(&sync_path, content)
            .map_err(|e| format!("write {}: {e}", sync_path.display()))?;
        files_written.push(sync_path.to_string_lossy().into_owned());

        let work_path = agents_dir.join("cortex-SDDwork.agent.md");
        backup_file(ctx, &work_path);
        let content = render_vscode_agent(
            &[
                "name: cortex-SDDwork",
                "description: Implement Cortex specs with fast-track or deep-track routing.",
                "tools: ['agent', 'edit', 'search/codebase', 'search/usages', 'cortex/*']",
                "agents: ['cortex-code-explorer', 'cortex-code-implementer', 'cortex-documenter']",
            ],
            &top_level_header,
            &work_body,
        );
        std::fs::write(&work_path, content)
            .map_err(|e| format!("write {}: {e}", work_path.display()))?;
        files_written.push(work_path.to_string_lossy().into_owned());

        for (agent_name, description, source) in &subagent_sources {
            let agent_header = generate_autogen_header(ctx, &[source], "vscode");
            let agent_path = claude_agents_dir.join(format!("{agent_name}.md"));
            backup_file(ctx, &agent_path);
            let canonical = get_subagent_prompt(ctx, agent_name);
            let content = render_claude_agent(
                agent_name,
                description,
                &agent_header,
                &strip_markdown_frontmatter(&canonical),
            );
            std::fs::write(&agent_path, content)
                .map_err(|e| format!("write {}: {e}", agent_path.display()))?;
            files_written.push(agent_path.to_string_lossy().into_owned());
        }

        Ok(files_written)
    }

    fn inject_mcp(&self, ctx: &IdeCtx) -> Result<Vec<String>, String> {
        let mcp_path = ctx.project_root.join(".vscode").join("mcp.json");
        std::fs::create_dir_all(mcp_path.parent().expect("mcp.json tiene parent"))
            .map_err(|e| format!("mkdir .vscode: {e}"))?;
        backup_file(ctx, &mcp_path);

        let mut data: Value = json!({});
        if mcp_path.exists() {
            if let Ok(text) = std::fs::read_to_string(&mcp_path) {
                if let Ok(parsed) = serde_json::from_str::<Value>(&text) {
                    data = parsed;
                }
            }
        }

        // VS Code usa ${workspaceFolder} (no ruta absoluta) para que el
        // workspace resuelva relativo al proyecto abierto.
        let cortex_config = json!({
            "type": "stdio",
            "command": "cortex-cli",
            "args": ["mcp-server", "--stdio", "--project-root", "${workspaceFolder}"],
            "env": {"PYTHONWARNINGS": "ignore"},
        });

        if let Some(obj) = data.as_object_mut() {
            obj.entry("servers".to_string())
                .or_insert_with(|| json!({}));
        }

        let merged = deep_merge_dict(
            data.get("servers").expect("servers setdefault"),
            &json!({"cortex": cortex_config}),
        );
        if let Some(obj) = data.as_object_mut() {
            obj.insert("servers".into(), merged);
        }

        std::fs::write(&mcp_path, json_dump_ascii(&data))
            .map_err(|e| format!("write {}: {e}", mcp_path.display()))?;
        Ok(vec![mcp_path.to_string_lossy().into_owned()])
    }

    /// Eliminar lo inyectado por Cortex en VS Code:
    ///
    /// - `.github/agents/{cortex-sync,cortex-SDDwork}.agent.md`.
    /// - `.claude/agents/{explorer,implementer,documenter}.md`.
    /// - Entrada `servers.cortex` de `<project>/.vscode/mcp.json`.
    ///
    /// SOLO se borra lo que Cortex creó (archivos por nombre canónico); los
    /// directorios solo se eliminan si quedan vacíos.
    fn uninstall(&self, ctx: &IdeCtx) -> Vec<String> {
        let mut removed: Vec<String> = Vec::new();
        let root = ctx.project_root;

        // (dir, sufijo, nombres canónicos) — orden del tuple-loop de Python.
        for (agents_dir, suffix, names) in [
            (
                root.join(".github").join("agents"),
                ".agent.md",
                &VSCODE_TOP_AGENTS[..],
            ),
            (
                root.join(".claude").join("agents"),
                ".md",
                &CLAUDE_SUBAGENTS[..],
            ),
        ] {
            for agent_name in names {
                let agent_path = agents_dir.join(format!("{agent_name}{suffix}"));
                if agent_path.exists() {
                    let _ = std::fs::remove_file(&agent_path);
                    removed.push(agent_path.to_string_lossy().into_owned());
                }
            }
            let is_empty = std::fs::read_dir(&agents_dir)
                .map(|mut d| d.next().is_none())
                .unwrap_or(false);
            if agents_dir.exists() && is_empty {
                let _ = std::fs::remove_dir(&agents_dir);
                removed.push(agents_dir.to_string_lossy().into_owned());
            }
        }

        // Entrada cortex de .vscode/mcp.json (merge inverso, preservando
        // otros servers del adopter). QUIRK de Python replicado: todo el
        // bloque vive dentro de contextlib.suppress(Exception), así que un
        // JSON inválido se salta SILENCIOSAMENTE (sin línea de reporte).
        let mcp_path = root.join(".vscode").join("mcp.json");
        if mcp_path.exists() {
            if let Ok(text) = std::fs::read_to_string(&mcp_path) {
                if let Ok(mut data) = serde_json::from_str::<Value>(&text) {
                    let mut should_write = false;
                    if let Some(obj) = data.as_object_mut() {
                        if let Some(Value::Object(servers)) = obj.get_mut("servers") {
                            // La remoción muta en lugar el map ya insertado
                            // en `data` (mismo efecto que el `del` de Python).
                            if servers.remove("cortex").is_some() {
                                should_write = true;
                            }
                        }
                    }
                    if should_write {
                        let _ = std::fs::write(&mcp_path, json_dump_ascii(&data));
                        removed.push(format!("{} (cortex entry removed)", mcp_path.display()));
                    }
                }
            }
        }

        removed
    }
}
